RFCs/rfc-0009-wallet-provider-store-request-api.md

```diff
@@ -0,0 +1,247 @@
+
+## Specification
+In the Web3 application ecosystem, which includes decentralized applications (dapps) and zero-knowledge applications (zkApps), it's a norm for key management software, known as "wallets," to make their API accessible through a JavaScript object in a web page—often termed "the Provider."
+
+However, inconsistencies have arisen in Provider implementations across wallets. This proposal seeks to standardize a Mina Provider API to ensure continued wallet interoperability with zkApps. This specification assumes wallets already implement the minimal, event-driven APIs equivalent to EIP-1193 for Mina; this specification introduces new RPC methods and message event types.
+
+Although Providers have historically been presented as `window.mina` in browsers, this convention isn't mandated in the specification.
+
+### API
+Wallets should have a provider API parallel to EIP-1193 but tailored for Mina. The goal of this API extension is to build on top of pre-existing standards. 
+
+#### Key Features of the Proposed API
+
+- **Agnosticity to Application Types**: Whether an application operates on-chain, off-chain, or a combination of the two, this API remains consistent and reliable. 
+
+- **Explicit User Control**: In a Mina-enabled DOM environment, the user is always in the driver's seat. The protocol is designed such that:
+
+    - **User Consent for Object Access**: Before a zkApp can access any object, explicit user approval is mandatory. 
+    - **User Consent for Storing Objects**: Just as access is restricted, storage of any objects by the wallet requires the user's green light. It ensures that users are aware of what's being stored in their wallets, enhancing transparency.
+
+#### Primary Functionalities:
+The proposed API should empower a Mina wallet to:
+- **Receive Objects**: Accept and store objects, whether they're verifiable credentials, proofs, or any other data.
+- **Request Objects**: Initiate requests to retrieve stored objects, facilitating interactions like transaction validations or data computations.
+
+### Scenario: A Wallet Receiving Objects to Store
+#### `mina_setState`
+Providers within Mina's domain could introduce a new RPC method: `mina_setState`. When called, a user interface might emerge, allowing users to choose whether to store objects from a zkApp. This method returns a Promise: resolving with a boolean or, if storing is unavailable, rejecting with an Error. A zkApp MUST provide sufficient information in the request to allow the wallet to store and index the object appropriately.
```

Comment:
What does it mean to "store objects from a zkApps"? I'm imagining that some zkApps would have public state - and others would store hashes of private state, and the private state is the thing that this is storing? Or something else?

Reply:
A zkApp can produce a well-formed credential that has a Kimchi proof in it, in that case, the zkApp would be the `issuer` of the credential and can allow the wallet to store that object for future use like in a smart contract method argument. 

If one were to assume that these objects could be fetched, or rather requested from a zkApp, this response can be stored much like how transaction history or current account information is stored in the wallet. The objects the wallet stores don't have to be private information but they can be.

A wallet storing credentials they can use in smart contract or provable-program methods is more about how to improve the interaction design between wallets and zkApps. If it is beneficial for a wallet to hold those credentials then they can use this API as a way of enabling that, there of course exist other designs that don't require the wallet to hold credentials but that design pattern is outside the scope of this RFC :)

Reply:
But how can users ensure these credentials are beneficial ?

Reply:
> But how can users ensure these credentials are beneficial ?

That is at up to the wallet user, the requested storing of those credentials must detail what the credential or object being stored is to the wallet user for them to determine the benefit. The real benefit of storing credentials in the wallet is that it gives the wallet user the ability to utilise them as and when they need/desire, without relying on central parties.